python/pyspark/sql/group.py

```diff
@@ -194,6 +194,28 @@ def pivot(self, pivot_col, values=None):
             jgd = self._jgd.pivot(pivot_col, values)
         return GroupedData(jgd, self.sql_ctx)
 
+    def apply(self, udf_obj):
+        """
+        Maps each group of the current [[DataFrame]] using a pandas udf and returns the result as a :class:`DataFrame`.
+
+        """
+        from pyspark.sql.functions import pandas_udf
+
+        df = DataFrame(self._jgd.df(), self.sql_ctx)
+        func = udf_obj.func
+        returnType = udf_obj.returnType
```

Comment:
Check if the return type is struct type?

Reply:
Done